src/analysis/calculus/taylor.lean

```diff
@@ -0,0 +1,380 @@
+    g g' gcont gdiff with ⟨y, hy, h⟩,
+  use [y, hy],
+  -- The rest is simplifications and trivial calculations
+  simp only [taylor_within_eval_self] at h,
+  rw [mul_comm, ←div_left_inj' (g'_ne y hy), mul_div_cancel _ (g'_ne y hy)] at h,
+  rw ←h,
+  field_simp [g'_ne y hy, nat.factorial_ne_zero n],
+  ring,
+end
+
+/-- **Taylor's theorem** with the Lagrange form of the remainder. -/
+lemma taylor_mean_remainder_lagrange {f : ℝ → ℝ} {x x₀ : ℝ} {n : ℕ} (hx : x₀ < x)
+  (hf : cont_diff_on ℝ n f (set.Icc x₀ x))
+  (hf' : differentiable_on ℝ (iterated_deriv_within n f (set.Icc x₀ x)) (set.Ioo x₀ x)) :
+  ∃ (x' : ℝ) (hx' : x' ∈ set.Ioo x₀ x), f x - taylor_within_eval f n (set.Icc x₀ x) x₀ x =
+  (iterated_deriv_within (n+1) f (set.Icc x₀ x) x') * (x - x₀)^(n+1) /(n+1).factorial :=
+begin
+  have gcont : continuous_on (λ (t : ℝ), (x - t) ^ (n + 1)) (set.Icc x₀ x) :=
+  by { refine continuous.continuous_on _, continuity },
+  have xy_ne : ∀ (y : ℝ), y ∈ set.Ioo x₀ x → (x - y)^n ≠ 0 :=
+  begin
+    intros y hy,
+    refine pow_ne_zero _ _,
+    rw [set.mem_Ioo] at hy,
+    rw sub_ne_zero,
+    exact hy.2.ne.symm,
+  end,
+  have hg' : ∀ (y : ℝ), y ∈ set.Ioo x₀ x → -(↑n + 1) * (x - y) ^ n ≠ 0 :=
+  λ y hy, mul_ne_zero (neg_ne_zero.mpr (nat.cast_add_one_ne_zero n)) (xy_ne y hy),
+  -- We apply the general theorem with g(t) = (x - t)^(n+1)
+  rcases taylor_mean_remainder hx hf hf' gcont (λ y _, monomial_has_deriv_aux y x) hg'
+    with ⟨y, hy, h⟩,
+  use [y, hy],
+  simp only [sub_self, zero_pow', ne.def, nat.succ_ne_zero, not_false_iff, zero_sub, mul_neg] at h,
+  rw [h, neg_div, ←div_neg, neg_mul, neg_neg],
+  field_simp [nat.cast_add_one_ne_zero n, nat.factorial_ne_zero n, xy_ne y hy],
+  ring,
+end
+
+/-- **Taylor's theorem** with the Cauchy form of the remainder. -/
+lemma taylor_mean_remainder_cauchy {f : ℝ → ℝ} {x x₀ : ℝ} {n : ℕ} (hx : x₀ < x)
+  (hf : cont_diff_on ℝ n f (set.Icc x₀ x))
+  (hf' : differentiable_on ℝ (iterated_deriv_within n f (set.Icc x₀ x)) (set.Ioo x₀ x)) :
+  ∃ (x' : ℝ) (hx' : x' ∈ set.Ioo x₀ x), f x - taylor_within_eval f n (set.Icc x₀ x) x₀ x =
+  (iterated_deriv_within (n+1) f (set.Icc x₀ x) x') * (x - x')^n /n.factorial * (x - x₀) :=
+begin
+  have gcont : continuous_on id (set.Icc x₀ x) := continuous.continuous_on (by continuity),
+  have gdiff : (∀ (x_1 : ℝ), x_1 ∈ set.Ioo x₀ x → has_deriv_at id
+    ((λ (t : ℝ), (1 : ℝ)) x_1) x_1) := λ _ _, has_deriv_at_id _,
+  -- We apply the general theorem with g = id
+  rcases taylor_mean_remainder hx hf hf' gcont gdiff (λ _ _, by simp) with ⟨y, hy, h⟩,
+  use [y, hy],
+  rw h,
+  field_simp [nat.factorial_ne_zero n],
+  ring,
+end
+
+/-- **Taylor's theorem** with a uniform bound of the remainder -/
+lemma taylor_mean_remainder_bound {f : ℝ → E} {a b : ℝ} {n : ℕ}
+  (h : a < b) (hf : cont_diff_on ℝ (n+1) f (set.Icc a b)) :
+  ∃ C : ℝ, ∀ (x : ℝ) (hx : x ∈ set.Icc a b),
+  ∥f x - taylor_within_eval f n (set.Icc a b) a x∥ ≤ C * (b - a)^(n+1) / n.factorial :=
```

Comment:
Could you write this one first with the upper bound `(x-a)^{n+1}/n!`, and then the version with `(b-a)^{n+1}` if you need it?

Reply:
I haven't gotten this to work with the version of the mean value theorem, I will try to figure out whether the mean value theorem is actually not strong enough or whether I missed some trick.

Reply:
Applying the current version with a=a and b=x should give you the stronger bound, right (upon checking that the C does not depend on x)? So your proof should already be strong enough.

Reply:
But the constant depends highly on `b`, and if we just set `b = x`, then the statement is trivially true. The issue I ran into was that you have to uniformly bound `|x-y|^n` and the only obvious bound is `(b-a)^n`. For the 1-d case the way out is to use a better mean value theorem in which you can put the higher powers of `x-y`, but this is simply not available for vector valued functions.
The result you want to have can be certainly proved with the integral form of the remainder, but it seems that integration by parts does not have a `smul` version yet.

Reply:
The bound is the sup norm of `f^(n)` on the interval `[a, b]`, so it works on all the intervals `[a, x]` as well for `x \in [a, b]`. I have pushed a proof (don't hesitate to change it as you like, and to extract the uniform version as well!)

Reply:
thanks for the commit, I was very stupid and tried way more complicated things that didn't work.